File: TempoCore/Tests/TempoCoreTests/ScheduleSpanTests.swift
// 템포루틴 — ScheduleSpan 테스트 (여러 날 일정, T80~)
// 주 경계에서 잘린 쪽은 각지게(isStart/isEnd = false), 실제 시작·끝만 둥글게 — 이 규칙을 못 박는다.

import XCTest
@testable import TempoCore

final class ScheduleSpanTests: XCTestCase {

    private var cal: Calendar {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = TimeZone(identifier: "Asia/Seoul") ?? .current
        return c
    }

    private func date(_ y: Int, _ m: Int, _ d: Int, hour: Int = 0, minute: Int = 0) -> Date {
        cal.date(from: DateComponents(year: y, month: m, day: d, hour: hour, minute: minute))!
    }

    /// 그리드 칸 포함 배열 만들기 — trueRange의 인덱스만 true
    private func cells(_ count: Int, _ trueIndices: [Int]) -> [Bool] {
        (0..<count).map { trueIndices.contains($0) }
    }

    // T80: 기간 계산 — 종료 없음·같은 날·여러 날·거꾸로
    func testT80_dayCount() {
        XCTAssertEqual(ScheduleSpan.dayCount(start: date(2026, 8, 3), end: nil, calendar: cal), 1)
        XCTAssertEqual(ScheduleSpan.dayCount(start: date(2026, 8, 3, hour: 9),
                                             end: date(2026, 8, 3, hour: 18), calendar: cal), 1)
        XCTAssertEqual(ScheduleSpan.dayCount(start: date(2026, 8, 3),
                                             end: date(2026, 8, 5), calendar: cal), 3)
        // 밤 11시 → 다음 날 새벽 1시 = 이틀에 걸침
        XCTAssertEqual(ScheduleSpan.dayCount(start: date(2026, 8, 3, hour: 23),
                                             end: date(2026, 8, 4, hour: 1), calendar: cal), 2)
        // 종료가 시작보다 이르면 1로 클램프
        XCTAssertEqual(ScheduleSpan.dayCount(start: date(2026, 8, 5),
                                             end: date(2026, 8, 3), calendar: cal), 1)
        // 월 경계
        XCTAssertEqual(ScheduleSpan.dayCount(start: date(2026, 7, 30),
                                             end: date(2026, 8, 2), calendar: cal), 4)
    }

    // T81: 한 주 안에서 이어지는 3일 → 조각 1개, 양 끝 둥글게
    func testT81_singleWeekRun() {
        let segments = ScheduleSpan.bandSegments(cells: cells(14, [8, 9, 10]))
        XCTAssertEqual(segments, [BandSegment(row: 1, column: 1, length: 3, isStart: true, isEnd: true)])
    }

    // T82: 주 경계를 넘으면 조각 2개 — 잘린 쪽은 각지게
    func testT82_weekBoundarySplit() {
        // 금(5)~월(8)
        let segments = ScheduleSpan.bandSegments(cells: cells(21, [5, 6, 7, 8]))
        XCTAssertEqual(segments, [
            BandSegment(row: 0, column: 5, length: 2, isStart: true, isEnd: false),
            BandSegment(row: 1, column: 0, length: 2, isStart: false, isEnd: true),
        ])
    }

    // T83: 3주에 걸치면 가운데 조각은 양쪽 다 각지게
    func testT83_threeWeeks() {
        let segments = ScheduleSpan.bandSegments(cells: cells(21, Array(6...15)))
        XCTAssertEqual(segments.count, 3)
        XCTAssertEqual(segments[0], BandSegment(row: 0, column: 6, length: 1, isStart: true, isEnd: false))
        XCTAssertEqual(segments[1], BandSegment(row: 1, column: 0, length: 7, isStart: false, isEnd: false))
        XCTAssertEqual(segments[2], BandSegment(row: 2, column: 0, length: 2, isStart: false, isEnd: true))
    }

    // T84: 달 밖에서 이어지면 그 끝은 둥글지 않다
    func testT84_monthBoundaryContinuation() {
        let head = ScheduleSpan.bandSegments(cells: cells(7, [0, 1]), continuesBefore: true)
        XCTAssertEqual(head, [BandSegment(row: 0, column: 0, length: 2, isStart: false, isEnd: true)])

        let tail = ScheduleSpan.bandSegments(cells: cells(7, [5, 6]), continuesAfter: true)
        XCTAssertEqual(tail, [BandSegment(row: 0, column: 5, length: 2, isStart: true, isEnd: false)])
    }

    // T85: 반복으로 끊긴 두 구간 → 각각 양 끝 둥글게
    func testT85_repeatedRunsAreSeparate() {
        let segments = ScheduleSpan.bandSegments(cells: cells(21, [1, 2, 8, 9]))
        XCTAssertEqual(segments, [
            BandSegment(row: 0, column: 1, length: 2, isStart: true, isEnd: true),
            BandSegment(row: 1, column: 1, length: 2, isStart: true, isEnd: true),
        ])
    }

    // T86: 하루짜리·빈 배열·전 구간
    func testT86_edges() {
        XCTAssertEqual(ScheduleSpan.bandSegments(cells: cells(7, [3])),
                       [BandSegment(row: 0, column: 3, length: 1, isStart: true, isEnd: true)])
        XCTAssertEqual(ScheduleSpan.bandSegments(cells: cells(7, [])), [])
        XCTAssertEqual(ScheduleSpan.bandSegments(cells: []), [])
        let full = ScheduleSpan.bandSegments(cells: cells(14, Array(0...13)))
        XCTAssertEqual(full.count, 2)
        XCTAssertEqual(full[0], BandSegment(row: 0, column: 0, length: 7, isStart: true, isEnd: false))
        XCTAssertEqual(full[1], BandSegment(row: 1, column: 0, length: 7, isStart: false, isEnd: true))
    }
}

File: android/tempocore/src/test/kotlin/app/temporoutine/core/ScheduleSpanTests.kt
// 템포루틴 — ScheduleSpan 테스트 (여러 날 일정, T80~)
// iOS TempoCoreTests/ScheduleSpanTests.swift 1:1 이식.

package app.temporoutine.core

import org.junit.jupiter.api.Test
import java.time.LocalDateTime
import kotlin.test.assertEquals

class ScheduleSpanTests {

    private fun date(y: Int, m: Int, d: Int, hour: Int = 0, minute: Int = 0): LocalDateTime =
        LocalDateTime.of(y, m, d, hour, minute)

    private fun cells(count: Int, trueIndices: List<Int>): List<Boolean> = (0 until count).map { it in trueIndices }

    // T80: 기간 계산
    @Test fun testT80_dayCount() {
        assertEquals(1, ScheduleSpan.dayCount(date(2026, 8, 3), null))
        assertEquals(1, ScheduleSpan.dayCount(date(2026, 8, 3, hour = 9), date(2026, 8, 3, hour = 18)))
        assertEquals(3, ScheduleSpan.dayCount(date(2026, 8, 3), date(2026, 8, 5)))
        assertEquals(2, ScheduleSpan.dayCount(date(2026, 8, 3, hour = 23), date(2026, 8, 4, hour = 1)))
        assertEquals(1, ScheduleSpan.dayCount(date(2026, 8, 5), date(2026, 8, 3)))
        assertEquals(4, ScheduleSpan.dayCount(date(2026, 7, 30), date(2026, 8, 2)))
    }

    // T81: 한 주 안에서 이어지는 3일
    @Test fun testT81_singleWeekRun() {
        assertEquals(listOf(BandSegment(1, 1, 3, true, true)), ScheduleSpan.bandSegments(cells(14, listOf(8, 9, 10))))
    }

    // T82: 주 경계를 넘으면 조각 2개
    @Test fun testT82_weekBoundarySplit() {
        assertEquals(listOf(
            BandSegment(0, 5, 2, true, false),
            BandSegment(1, 0, 2, false, true),
        ), ScheduleSpan.bandSegments(cells(21, listOf(5, 6, 7, 8))))
    }

    // T83: 3주에 걸치면 가운데 조각은 양쪽 다 각지게
    @Test fun testT83_threeWeeks() {
        val segments = ScheduleSpan.bandSegments(cells(21, (6..15).toList()))
        assertEquals(3, segments.size)
        assertEquals(BandSegment(0, 6, 1, true, false), segments[0])
        assertEquals(BandSegment(1, 0, 7, false, false), segments[1])
        assertEquals(BandSegment(2, 0, 2, false, true), segments[2])
    }

    // T84: 달 밖에서 이어지면 그 끝은 둥글지 않다
    @Test fun testT84_monthBoundaryContinuation() {
        assertEquals(listOf(BandSegment(0, 0, 2, false, true)),
            ScheduleSpan.bandSegments(cells(7, listOf(0, 1)), continuesBefore = true))
        assertEquals(listOf(BandSegment(0, 5, 2, true, false)),
            ScheduleSpan.bandSegments(cells(7, listOf(5, 6)), continuesAfter = true))
    }

    // T85: 반복으로 끊긴 두 구간
    @Test fun testT85_repeatedRunsAreSeparate() {
        assertEquals(listOf(
            BandSegment(0, 1, 2, true, true),
            BandSegment(1, 1, 2, true, true),
        ), ScheduleSpan.bandSegments(cells(21, listOf(1, 2, 8, 9))))
    }

    // T86: 하루짜리·빈 배열·전 구간
    @Test fun testT86_edges() {
        assertEquals(listOf(BandSegment(0, 3, 1, true, true)), ScheduleSpan.bandSegments(cells(7, listOf(3))))
        assertEquals(emptyList(), ScheduleSpan.bandSegments(cells(7, emptyList())))
        assertEquals(emptyList(), ScheduleSpan.bandSegments(emptyList()))
        val full = ScheduleSpan.bandSegments(cells(14, (0..13).toList()))
        assertEquals(2, full.size)
        assertEquals(BandSegment(0, 0, 7, true, false), full[0])
        assertEquals(BandSegment(1, 0, 7, false, true), full[1])
    }
}
